Sources/SwiftBSON/BSONDocument.swift

```diff
@@ -517,3 +517,42 @@ extension BSONDocument: BSONValue {
 extension BSONDocument: CustomStringConvertible {
     public var description: String { self.toExtendedJSONString() }
 }
+
+extension BSONDocument {
+    /// Helper function to check for BSON document equality regardless of the order of key-value pairs
+    internal func equalsIgnoreOrderHelper(a: BSONDocument, b: BSONDocument) -> Bool {
+        let keysA = a.keys.sorted()
+        let keysB = b.keys.sorted()
+
+        guard keysA == keysB else {
+            return false
+        }
+
+        for k in keysA {
+            if case let (.document(aDoc), .document(bDoc)?) = (a[k], b[k]) {
+                guard self.equalsIgnoreOrderHelper(a: aDoc, b: bDoc) else {
+                    return false
+                }
+                continue
+            }
+            guard a[k] == b[k] else {
+                return false
+            }
+        }
+
+        return true
+    }
+
+    /**
+     * Compares this `BSONDocument` with another provided one and returns true if they have the same key/value pairs
+     * in them regardless of the order.
+     *
+     * - Parameters:
+     *   - other: a `BSONDocument` to compare this document with.
+     *
+     * - Returns: a `Bool` indicating whether the two documents are equal.
+     */
+    public func equalsIgnoreOrder(_ other: BSONDocument) -> Bool {
```

Comment:
I think we may need to specify "KeyOrder", since we do want BSON arrays to be compared in the proper order.